File: library/server/src/main/java/cz/matfyz/server/entity/job/data/InferenceJobData.java
package cz.matfyz.server.entity.job.data;

import cz.matfyz.core.mapping.Mapping;
import cz.matfyz.core.mapping.Mapping.SerializedMapping;
import cz.matfyz.core.metadata.MetadataCategory;
import cz.matfyz.core.metadata.MetadataSerializer;
import cz.matfyz.core.metadata.MetadataSerializer.SerializedMetadata;
import cz.matfyz.core.rsd.Candidates;
import cz.matfyz.core.rsd.CandidatesSerializer;
import cz.matfyz.core.rsd.CandidatesSerializer.SerializedCandidates;
import cz.matfyz.core.schema.SchemaCategory;
import cz.matfyz.core.schema.SchemaSerializer;
import cz.matfyz.core.schema.SchemaSerializer.SerializedSchema;
import cz.matfyz.inference.edit.InferenceEdit;
import cz.matfyz.inference.schemaconversion.utils.LayoutType;
import cz.matfyz.server.entity.job.JobData;

import java.util.ArrayList;
import java.util.List;
import java.util.TreeMap;

public record InferenceJobData(
    List<InferenceEdit> edits,
    SerializedSchema inferenceSchema,
    SerializedSchema finalSchema,
    SerializedMetadata inferenceMetadata,
    SerializedMetadata finalMetadata,
    LayoutType layoutType,
    SerializedCandidates candidates,
    List<SerializedDatasource> datasources
) implements JobData {

    public static InferenceJobData fromSchemaCategory(
        List<InferenceEdit> edits,
        SchemaCategory inferenceSchema,
        SchemaCategory finalSchema,
        MetadataCategory inferenceMetadata,
        MetadataCategory finalMetadata,
        LayoutType layoutType,
        Candidates candidates,
        List<Mapping> mappings
    ) {
        final var datasources = new TreeMap<String, SerializedDatasource>();
        mappings.forEach(mapping -> {
            final var datasourceId = mapping.datasource().identifier;
            datasources
                .computeIfAbsent(datasourceId, id -> new SerializedDatasource(id, new ArrayList<>()))
                .mappings().add(SerializedMapping.fromMapping(mapping));
        });

        return new InferenceJobData(
            edits,
            SchemaSerializer.serialize(inferenceSchema),
            SchemaSerializer.serialize(finalSchema),
            MetadataSerializer.serialize(inferenceMetadata),
            MetadataSerializer.serialize(finalMetadata),
            layoutType,
            CandidatesSerializer.serialize(candidates),
            datasources.values().stream().toList()
        );
    }

    public record SerializedDatasource(
        String datasourceId,
        List<SerializedMapping> mappings
    ) {}

}

File: client-old/src/types/inference/InferenceJobData.ts
import type { JobDataType } from '../job';
import { Category, type SchemaCategoryInfo, type SerializedSchema, type SerializedMetadata } from '../schema';
import { Candidates, type SerializedCandidates } from './candidates';
import type { InferenceEdit, SerializedInferenceEdit } from './inferenceEdit';
import { createInferenceEditFromServer } from './inferenceEdit';
import type { LayoutType } from './layoutType';

/**
 * Type representing the data structure of an inference job received from the server.
 */
export type InferenceJobDataFromServer = {
    /** The type of job, in this case, an inference job. */
    type: JobDataType.Inference;
    /** The serialized inference edits associated with the job. */
    edits: SerializedInferenceEdit[];
    /** The serialized inference schema. */
    inferenceSchema: SerializedSchema;
    /** The serialized final schema. */
    finalSchema: SerializedSchema;
    /** Metadata related to the inference schema. */
    inferenceMetadata: SerializedMetadata;
    /** Metadata related to the final schema. */
    finalMetadata: SerializedMetadata;
    /** The layout type used for visualizing the data. */
    layoutType: LayoutType;
    /** Serialized candidates data for the job. */
    candidates: SerializedCandidates;
};

/**
 * Class representing the data for an inference job.
 */
export class InferenceJobData {
    constructor(
        public edits: InferenceEdit[],
        public inferenceSchema: Category,
        public finalSchema: Category,
        public layoutType: LayoutType,
        public candidates: Candidates,
    ) {}

    /**
     * Creates an instance of `InferenceJobData` from server data.
     */
    static fromServer(input: InferenceJobDataFromServer, info: SchemaCategoryInfo): InferenceJobData {
        return new InferenceJobData(
            input.edits.map(createInferenceEditFromServer),
            Category.fromServerWithInfo(info, input.inferenceSchema, input.inferenceMetadata),
            Category.fromServerWithInfo(info, input.finalSchema, input.finalMetadata),
            input.layoutType,
            Candidates.fromServer(input.candidates),
        );
    }
}
